torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+import functools
+import inspect
+import warnings
+from typing import Callable, Optional, Tuple
+
+import torch
+from torch.testing._internal.common_utils import get_comparison_dtype as _get_comparison_dtype, TestCase as _TestCase
+
+__all__ = ["assert_tensors_equal", "assert_tensors_allclose"]
+
+try:
+    import pytest
+
+    UsageError = pytest.UsageError
+except ImportError:
+
+    class UsageError(Exception):
+        pass
+
+
+class UsageWarning(Warning):
+    pass
```

Comment:
We add a custom error / warning in case someone is using the functions wrong. With this we can differentiate between errors in the code and in the test suite.

Reply:
I am not convinced this is worth importing pytest. The only difference when using pytest.UsageError under pytest is that the exit value for a usage error is changed to `USAGE_ERROR` (4). On the other hand, avoiding pytest altogether here signals that these functions are not test-package specific.

Reply:
> I am not convinced this is worth importing pytest. The only difference when using pytest.UsageError under pytest is that the exit value for a usage error is changed to `USAGE_ERROR` (4).

I can live with a `UsageError` without importing `pytest`. @mruberry?

> On the other hand, avoiding pytest altogether here signals that these functions are not test-package specific.

Do you suspect the users will look at the source, see this line and deduce that they need `pytest`? IMO, if we don't mention any test frameworks in the documentation, it should be clear, that these functions are framework agnostic.

Reply:
Both positions seem reasonable but I'd remove it just to reduce the maintenance burden if it's not a clear win. We can remove it in a follow-up PR, however. It doesn't need to block this one.